Design.md

```diff
@@ -0,0 +1,241 @@
+  + [Create User Example Activity Diagram](#create-user-example-activity-diagram)
+  + [Query Users](#query-users)
+  + [List Users](#list-users)
+  + [Get User w/ userid](#get-user-w--userid)
+  + [Get User w/ canonical ID](#get-user-w--canonical-id)
+  + [Head User](#head-user)
+  + [Delete User](#delete-user)
+  + [Update User](#update-user)
+* [S3 Credential APIs](#s3-credential-apis)
+  + [Create S3 Credential](#create-s3-credential)
+  + [Query S3 Credentials](#query-s3-credentials)
+  + [List S3 Credentials](#list-s3-credentials)
+  + [Delete S3 Credential](#delete-s3-credential)
+  + [Get S3 Credential](#get-s3-credential)
+* [Miscellaneous APIs:](#miscellaneous-apis-)
+  + [getProviderConsoleUrl](#getproviderconsoleurl)
+  + [getTenantConsoleUrl](#gettenantconsoleurl)
+  + [getS3Capabilities](#gets3capabilities)
+  + [getInformation](#getinformation)
+  + [updateOsisCaps (?)](#updateosiscaps----)
+  + [getBucketList](#getbucketlist)
+  + [getOsisUsage](#getosisusage)
+* [S3 Operations Example Activity Diagram](#s3-operations-example-activity-diagram)
+</details>
+
+## OSIS inital setup
+
+For the initial installation of OSIS application, following steps must be executed.
+1. A **Super Admin** account will be created as part of S3C/Vault setup
+1. Set OSE properties:
+   1. S3 URL
+   1. Admin Vault URL along with credentials as Super Admin access key/ secret key
+    
+_**Note: "Super Admin" account has full Account Management access and can use "AssumeRoleBackbeat" to assume a role of another Account.**_
+
+## Tenant APIs
+<u>**Important Assumptions:**</u>
+* All the Tenant APIs on OSIS will use **Super Admin credentials** to manage on Vault.
+* A Vault Account is equivalent of OSE Tenant.
+
+### Create Tenant
+This API creates a Tenant on Vault.  
+1. Vault `create-account` api will be called using vaultclient. 
+    * Tenant Name is stored as Account Name  
+    * `cd_tenant_ids` list as part of the request to be stored as a property of the vault account.
+    * Account ID to be sent back as storage tenant ID to OSE
+1. `generate-account-access-key` for the account using vaultclient.
+1. Create a role with
+   1. Role name : `osis` (So role-arn can be generated in the format `arn:aws:iam::[account-id]:role/osis`
+   1. Trust policy
+   ```json
+    {
+        "Version": "2012-10-17",
+        "Statement": [{
+            "Effect": "Allow",
+            "Principal": {
+                "Service": "service-name"
+            },
+            "Action": "sts:AssumeRole"
+            }]
+    }
+   ```
+1. `attach-role-policy` with s3 and iam full access to this role
+1. `delete-access-key` for the account.
+
+### Create Tenant Example Activity Diagram
+![Create Tenant](Design_Files/OSE-OSIS-Create-Tenant.png)
+
+### Query Tenants
+This API will query tenants on Vault using a `filter` parameter.
+1. Vault `list-accounts` api will be called using vaultclient.
+1. Filter the result accounts using the `filter` parameter in the request
```

Comment:
Worth mentioning which service will filder the accounts.
We also need to document how will `list-accounts` be called again, if there are more than 1000 accounts.
Will, there be an infinite call with each marker, till we have `isTrucated` not true in the result? If yes, then we might to document that.

Reply:
> if there are more than 1000 accounts.
> Will, there be an infinite call with each marker, till we have `isTrucated` not true in the result? If yes, then we might to document that.

yes this is what expected with respect to marker if no limit field is sent. 
List Tenants use `limit` and `offset` optional parameters
`offset`: The start index of tenants to return (optional) => equivalent to `marker`
`limit`:  Maximum number of tenants to return (optional) => equivalent to `max-items`

Reply:
Updated in commit [59a5a88](https://github.com/scality/vmware-ose-scality/pull/11/commits/59a5a884aa427cfb91917ac2bc196ee07651ebd8)